src/polyfillstub/intersection-observer-stub.js

```diff
@@ -60,9 +73,25 @@ export function upgradePolyfill(win, installer) {
   // instance deployed in v0.js vs the polyfill extension.
   const Stub = /** @type {typeof IntersectionObserverStub} */ (win.IntersectionObserver);
   if (Stub && UPGRADERS in Stub) {
+    const NativeIntersectionObserver =
+      win.IntersectionObserver && win.IntersectionObserver.native;
     delete win.IntersectionObserver;
     delete win.IntersectionObserverEntry;
     installer();
+    const PolyfilledIntersectionObserver = win.IntersectionObserver;
+    // NativeIntersectionObserver only exists in the case of an InOb
+    // that doesn't support {root: document}. In this case we'd like to
+    // select the fastest InOb we can on a per-instance basis.
+    if (NativeIntersectionObserver) {
+      win.IntersectionObserver = function (ioCallback, opts) {
+        if (opts && opts.root && opts.root.nodeType === 9) {
+          return new PolyfilledIntersectionObserver(ioCallback, opts);
+        } else {
+          return new NativeIntersectionObserver(ioCallback, opts);
```

Comment:
If I'm reading it right, the `NativeIntersectionObserver` can be null.

Reply:
how can it be null? I have an if-check for truthy `NativeIntersectionObserver` immediately above!

Reply:
Sorry, my mistake. I was still reacting to [this thread](https://github.com/ampproject/amphtml/pull/31453/files#r537771502). Let's close this one and continue discussion there. But otherwise, correct - no problem here. Though if you keep it this way, I recommend you move `const PolyfilledIntersectionObserver = ...` inside the `if (NativeInOb)`.